src/components/card/card.component.ts

```diff
@@ -1,26 +1,48 @@
 import { Component, Input, OnInit } from '@angular/core';
 import { humanizeMeasureName } from '../../utils/formatters';
-
-export type CardType = 'line' | 'bar';
+import { Meta } from '../../datasets/metas/types';
+import { TabbedChartsMeta } from '../../datasets/metas/tabbed-charts.meta';
 
 @Component({
   selector: 'app-card',
   templateUrl: './card.component.html',
   styleUrls: ['./card.component.scss'],
 })
-export class CardComponent implements OnInit {
-  @Input() title: string;
-  @Input() tabbed = false;
-  @Input() type: CardType;
-  @Input() measureNames: string[];
+export class CardComponent<T extends Meta> implements OnInit {
+  @Input() meta: T;
   humanizeMeasureName = humanizeMeasureName;
-  currentMeasureName: string;
+  currentTabTitle: string;
+
+  get titles() {
+    if (this.isTabbed()) {
+      return this.meta.metas.map(c => c.title);
+    } else {
+      return [this.meta.title];
+    }
+  }
+
+  get currentChart(): Meta {
+    if (this.isTabbed()) {
+      const { metas } = this.meta;
+      return metas.find(c => c.title === this.currentTabTitle) ?? metas[0];
+    } else {
+      return this.meta;
+    }
+  }
 
   ngOnInit() {
-    this.setMeasureName(this.measureNames[0]);
+    if (this.isTabbed()) {
+      this.setCurrentTabTitle(this.meta.metas[0].title);
+    } else {
+      this.setCurrentTabTitle('');
+    }
+  }
+
+  setCurrentTabTitle(title) {
+    this.currentTabTitle = title;
   }
 
-  setMeasureName(measureName) {
-    this.currentMeasureName = measureName;
+  isTabbed(): this is CardComponent<TabbedChartsMeta> {
```

Comment:
Can this type just be boolean?

Reply:
If it were `boolean` type instead of the type guard, inside the if-statements (line 18, 26, 35) would need unnecessary type assertions like `return (this.meta as TabbedChartsMeta).metas.map(c => c.title);`.

Reply:
Oh I see, I did not make that connection. Cool but hidden type guard!

Could this be a bit cleaner if the CardComponent could take in an input of chartMeta arrays or something like a CardMeta instead of any Meta. The CardMeta could just have an array of ChartMetas and any other card level information. This way you don't need to narrow from Meta to TabbedChartsMeta. And to avoid the split logic for many of the get methods, could we just keep a reference to the current chart meta?

lmk what you think

Reply:
Sounds good, does it mean `ChartMeta` will always be wrapped in some kinds of `CardMeta`?
When designing this part, I assume that we can have nested `CardMeta`, like a grid card with one of block being a tabbed card (charts). In this case, we still need to have some logic inside card component to narrow from `Meta` to `ChartMeta` or CardMeta. A better practice is to make these card into different CardComponents, like `TabbedCardComponent` and `SingleChartCardComponent`, but it will be lots of refactoring works and a separate PR.
I dont know if my assumption and implementation is over engineering. Maybe we just need some very simple card structure.

Reply:
I think it's okay to leave it as is! Perhaps add a comment explaining that this is a type guard as well!

Reply:
Something like "Type guard for narrowing down meta type to "TabbedChartsMeta""